lib/mix/lib/mix/cli.ex

```diff
@@ -104,23 +104,40 @@ defmodule Mix.CLI do
     end
   end
 
-  defp change_env(task) do
-    if env = preferred_cli_env(task) do
-      Mix.env(env)
+  defp maybe_change_env_and_target(task) do
+    task = String.to_atom(task)
+    config = Mix.Project.config()
 
-      if project = Mix.Project.pop() do
-        %{name: name, file: file} = project
-        Mix.Project.push(name, file)
-      end
+    env = preferred_cli_env(task, config)
+    target = preferred_cli_target(task, config)
+    env && Mix.env(env)
+    target && Mix.target(target)
+
+    if env || target do
+      reload_project()
     end
   end
 
-  defp preferred_cli_env(task) do
+  defp reload_project() do
+    if project = Mix.Project.pop() do
+      %{name: name, file: file} = project
+      Mix.Project.push(name, file)
+    end
+  end
+
+  defp preferred_cli_env(task, config) do
     if System.get_env("MIX_ENV") do
       nil
     else
-      task = String.to_atom(task)
-      Mix.Project.config()[:preferred_cli_env][task] || Mix.Task.preferred_cli_env(task)
+      config[:preferred_cli_env][task] || Mix.Task.preferred_cli_env(task)
+    end
+  end
+
+  defp preferred_cli_target(task, config) do
+    if System.get_env("MIX_TARGET") do
```

Comment:
This no longer functions in a desirable way. Since MIX_TARGET is almost always set when cross compiling, this code almost always has no effect. I this condition needs to be eliminated and it should always check the config and task.

Reply:
Hrm, this means then there is no way for me to override a default imposed by the task, and I think we ultimately need to provide a way for the developer to override that.

Why are you setting MIX_TARGET by hand? I understand that originally we had to do it to support both old and new Elixir versions alike... but maybe you can always set the target via `Mix.target` and leave the MIX_TARGET environment variable for only when users set it?

Reply:
The user almost always has `MIX_TARGET` exported when working on a project that is cross compiled since they are spending a majority of their time compiling things for the target. Otherwise they would be running `MIX_TARGET=rpi3 mix compile` `MIX_TARGET=rpi3 mix firmware` `MIX_TARGET=rpi3 mix burn` all the time. The exception comes in for `mix run` and `mix test` since these are both situations where the execution should be for the host.

If the user wants to override the behavior I suggested above, they could set it to the original target in the mix file. `preferred_cli_target: [test: Mix.target()]` But I would argue that this would be highly unlikely in the case of `mix test`

If adding `@preferred_cli_target` to mix test feels too controversial, then we could add it to the mix file as a suggested default.